src/libexpr/primops.cc

```diff
@@ -1621,7 +1621,13 @@ static RegisterPrimOp primop_toJSON({
 static void prim_fromJSON(EvalState & state, const Pos & pos, Value * * args, Value & v)
 {
     string s = state.forceStringNoCtx(*args[0], pos);
-    parseJSON(state, s, v);
+    try {
+        parseJSON(state, s, v);
+    } catch (JSONParseError &e) {
+        auto info = e.info();
+        info.errPos = pos;
+        throw EvalError(info);
```

Comment:
Isn't it better to add the error position to `e` (maybe using `addErrorTrace`) and then rethrow `e`? That way it doesn't change the type.